polkadot/xcm/pallet-xcm/src/lib.rs

```diff
@@ -2421,17 +2422,16 @@ impl<T: Config> Pallet<T> {
 		mut message: Xcm<()>,
 	) -> Result<XcmHash, SendError> {
 		let interior = interior.into();
+		let origin_dest = interior.clone().into();
 		let dest = dest.into();
-		let maybe_fee_payer = if interior != Junctions::Here {
-			message.0.insert(0, DescendOrigin(interior.clone()));
-			Some(interior.into())
-		} else {
-			None
-		};
+		let is_waived = <T::XcmExecutor as FeeManager>::is_waived(Some(&origin_dest), FeeReason::ChargeFees);
+		if !is_waived {
+			message.0.insert(0, DescendOrigin(interior));
+		}
```

Comment:
This is not really what we want.

Whether or not `DescendOrigin(interior)` is inserted is orthogonal to whether fees are paid; it should depend on the same condition as before:

```suggestion
		if interior != Junctions::Here {
			message.0.insert(0, DescendOrigin(interior));
		}
```